[FLINK-15794][Kubernetes] Generate the Kubernetes default image version

nielsbasjes opened the pull request:
## What is the purpose of the change

The default image used by Kubernetes is `flink:latest` which causes version compatibility problems if the latest it not exactly the same as what you are using.
This change generates the exact version id of the image (i.e. no longer `latest`).

## Brief change log

- Generate the project and scala versions into a simple Java class.
- Use this generated code to define the docker image tag and related documentation.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

This changes only the default value.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no, only build plugins
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes, by default a different image tag is pulled.
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs

wangyang0918 commented on docs/ops/deployment/kubernetes.md:
```diff
@@ -112,7 +112,7 @@ An early version of a [Flink Helm chart](https://github.com/docker-flink/example
 ### Session cluster resource definitions
 
-The Deployment definitions use the pre-built image `flink:latest` which can be found [on Docker Hub](https://hub.docker.com/r/_/flink/).
+The Deployment definitions use the pre-built image `flink:{{ site.version }}-scala_2.11` (or `flink:{{ site.version }}-scala_2.12`) which can be found [on Docker Hub](https://hub.docker.com/r/_/flink/).
```
I am afraid we could not use `{{ site.version }}` here. We have at lease two reasons.
* When the minor version increase, the `{{ site.version }}` do not change. For example, 1.9.2 has released, and the `{{ site.version }}` is still 1.9.0.
* We do not have a image version for the master branch document.

nielsbasjes replied:
- Strange. I see here that the site.version for 1.9.2 was actually set to 1.9.2 : https://github.com/apache/flink/blob/release-1.9.2/docs/_config.yml#L30 . I do agree that for example https://ci.apache.org/projects/flink/flink-docs-release-1.9/dev/connectors/pubsub.html still shows 1.9.0. To me this seems more like the site needs to be regenerated than this variable having the incorrect value.

- Perhaps on releasing a 'SNAPSHOT' version of the software a 'SNAPSHOT' version docker image with the same software should be released too?

On the other hand ... it is 'just documentation'.
I do see that apparently the 2.12 is 'preferred' so I'm swapping them in the documentation.

wangyang0918 replied:
Releasing a `SNAPSHOT` image tag is still in discussion. Currently, Flink only provide the image for released versions.

azagrebin replied:
As I mentioned this change is not strictly related to the JIRA at hand, as it is about standalone kubernetes deployment and not about the native kubernetes image option.
It seems this `site.version` is not reliable at the moment. I agree with @wangyang0918,
this can bring a confusion.
I still think we should create another issue for this doc change which depends on 2 problems to be resolved: 
- the image Flink version should point to the latest minor version,
- there are no Flink snapshot versions in docker hub at the moment so master doc will point to a not existing image.

azagrebin replied:
We had the following idea with @zentol (you can check it in #12131):
```
flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %}
```

This way we keep it a concrete version for the stable released docs, like e.g. `1.10.1` (the bug fix release guide should include incrementing minor version in `site.version`, this should change after `1.11`). There is always a docker image for the released versions.

The master docs (not released, like 1.11-SNAPHOST) will always point to the `latest` released docker image for consistency for now until we introduce snapshots for docker image as well.

I know that the implicit scala version is currently `2.12` for short docker tags but it is `2.11` for everything else. `site.scala_version` is also `2.11`. It should have probably been `2.11` for docker tags as well.

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -121,8 +122,11 @@
 		.stringType()
-		.defaultValue("flink:latest")
-		.withDescription("Image to use for Flink containers.");
+		.defaultValue("flink:" + Version.PROJECT_VERSION + "-scala_" + Version.SCALA_VERSION)
```
Could we simply use `EnvironmentInformation#getVersion()` to get the project version? And i think we do not need the scalar version. Since we always have a image tag corresponding to the Flink version(with scala 2.12 built in).

nielsbasjes replied:
I reported https://issues.apache.org/jira/browse/FLINK-16289 last week about getting java.io.InvalidClassException: org.apache.flink.table.codegen.GeneratedAggregationsFunction.
Root cause after several days of digging:
- I had scala 2.11 on my machine
- Downloaded Flink 1.10.0 for scala 2.11 to start my session on Kubernetes.
- Used Flink 1.10.0 for scala 2.11 for all dependencies in my project that I built and ran.
- Then "Flink 1.10.0 for scala 2.11" on Kubernetes downloaded "flink:latest" image which is really "Flink 1.10.0 for scala 2.12" which broke it all.

Also if for some reason I choose not to upgrade and a new release for Flink is created (i.e. 1.11) then my application may also break.

So I think "Flink X for scala Y" must download the docker image for exactly THAT combination of versions.

So I think having the scala version in there explicitly is a must have.

or ... We ditch the scala 2.11 variant completely and have only 1 scala variant.

I'll have a look at the EnvironmentInformation you mentioned, perhaps that is the better place to add the scala version.

nielsbasjes commented on docs/ops/deployment/kubernetes.md:
```diff
@@ -112,7 +112,7 @@ An early version of a [Flink Helm chart](https://github.com/docker-flink/example
 
-The Deployment definitions use the pre-built image `flink:latest` which can be found [on Docker Hub](https://hub.docker.com/r/_/flink/).
+The Deployment definitions use the pre-built image `flink:{{ site.version }}-scala_2.11` (or `flink:{{ site.version }}-scala_2.12`) which can be found [on Docker Hub](https://hub.docker.com/r/_/flink/).
 The image is built from this [Github repository](https://github.com/docker-flink/docker-flink).
```
Yes, I'll update that.

nielsbasjes commented:
@wangyang0918 I have moved the version things towards the EnvironmentInformation class.
Because I could not get to the used scala version via the method that was used there I refactored that part to be consistent.
Please review.

nielsbasjes commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,14 +44,19 @@
 	public static String getVersion() {
-		String version = EnvironmentInformation.class.getPackage().getImplementationVersion();
-		return version != null ? version : UNKNOWN;
+		return PROJECT_VERSION;
```
The `class.getPackage().getImplementationVersion();` is a very complex way of finding the version of the project at hand. Simply put: Get the same value as is available in the project.version property in the pom.xml. 
I had to generate a Version.java file with the used scala version in it because I was unable to get that property via a similar route. So for code maintainability I simplified it all and use the same system to get both values.

nielsbasjes commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -41,14 +44,19 @@
+	 * @return The scala version string.
+	 */
+	public static String getScalaVersion() {
+		return SCALA_VERSION;
```
I'll have a look. I didn't know about this method.

nielsbasjes replied:
I checked and on my machine this returns "2.11.12" which is different from the expected "2.11".

azagrebin replied:
yes but we could easily trim it, right?

nielsbasjes replied:
Yes, but is this the correct version in all cases? What do you get if Flink is built against 2.11 and you run it with the 2.12 runtime?
I really prefer the generation of the build time versions in the code in a predictable way.

azagrebin replied:
Indeed, I do also see any other way which would be simple but reliable.

nielsbasjes commented on flink-runtime/pom.xml:
```diff
@@ -475,6 +542,7 @@ under the License.
 						<!-- don't generate the describe property -->
 						<skip>true</skip>
 					</gitDescribe>
+					<dateFormat>yyyy-MM-dd '@' HH:mm:ss z</dateFormat>
```
The default version format of the git information plugin is a (very confusing) local format that is only used in this way in the US. The output of this plugin is put in all logging and in the webui as-is.

nielsbasjes commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -121,8 +122,11 @@
 		.stringType()
-		.defaultValue("flink:latest")
-		.withDescription("Image to use for Flink containers.");
+		.defaultValue("flink:" + EnvironmentInformation.getVersion() + "-scala_" + EnvironmentInformation.getScalaVersion())
```
Ok

azagrebin replied:
Do you think now we should not put this to a private helper method/static var?

nielsbasjes commented on docs/ops/deployment/kubernetes.md:
```diff
@@ -112,8 +112,8 @@ An early version of a [Flink Helm chart](https://github.com/docker-flink/example
 
-The Deployment definitions use the pre-built image `flink:latest` which can be found [on Docker Hub](https://hub.docker.com/r/_/flink/).
-The image is built from this [Github repository](https://github.com/docker-flink/docker-flink).
+The Deployment definitions use the pre-built image `flink:{{ site.version }}-scala_2.12` (or `flink:{{ site.version }}-scala_2.11`) which can be found [on Docker Hub](https://hub.docker.com/_/flink/).
```
As a rule for myself I always try to make the documentation match the change I'm writing.
In this case for me this is really about avoiding the extremely obscure errors you get if you use the wrong Scala version docker image. So in this case to me this documentation seemed like it needed to be fixed as well as part of this change set.

Side note:
Given that the flink docker image must match both the Flink version and Scala version used by my application to run successfully: I would vote to completely drop all docker tags that do not explicitly specify both Flink and Scala versions (i.e. no more `latest` , `scala_2.12` or `1.10.0` tags).

wangyang0918 replied:
I agree with @azagrebin that we should move out this changes. First, it is not related to this ticket(default image version of native K8s). Second, we could not guarantee that it is always existing in the docker hub.

nielsbasjes commented on flink-runtime/pom.xml:
```diff
@@ -402,6 +402,59 @@ under the License.
 			</plugin>
 
+			<plugin>
+				<groupId>com.google.code.maven-replacer-plugin</groupId>
```
To me this is the cleanest way to get the correct values needed.

azagrebin replied:
cc @zentol

azagrebin replied:
We synced offline with @zentol. In general, we agree that this is a cleaner way to go as we explicitly use versions defined in one place. Nonetheless, as I mentioned, this is a separate issue which could be a prerequisite for the change at hand (default value of Flink image for native k8s integration). There are some open questions to resolve in that separate issue.  E.g. the setup of most IDE projects  will be broken w/o code generation as `Version` class is not committed, so this is going to be annoying. If we commit it, it can easily become stale then it is better to keep then "unknown" for IDE builds and generate into the actual built artefacts. Committing the actual values all the time is also annoying, especially for the git data.

nielsbasjes replied:
OK, I'll create a separate issue for generating the Versions file. Note that in my experience (IntelliJ) has a generate sources button in the area of maven which simply regenerates (among other things) this file. This is exactly the same pattern as used with all other generated code; for example Antlr4.

nielsbasjes replied:
I created https://issues.apache.org/jira/browse/FLINK-16871 . Would you please assign this to me? Thanks.

nielsbasjes commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -121,8 +122,11 @@
-		.withDescription("Image to use for Flink containers.");
+		.defaultValue("flink:" + EnvironmentInformation.getVersion() + "-scala_" + EnvironmentInformation.getScalaVersion())
+		.withDescription("Image to use for Flink containers. " +
+			"The specified image MUST be based upon Apache Flink " +
```
What are the guidelines on compatibility between different versions?

azagrebin replied:
What is the conclusion here?

nielsbasjes commented:
I did a rebase against the current master branch and moved the docker image name into a separate function.

nielsbasjes commented:
As discussed I created https://issues.apache.org/jira/browse/FLINK-16871 and the the relevant changes in https://github.com/apache/flink/pull/11592 .
Can you guys please assign the Jira ticket to me?

azagrebin commented:
@nielsbasjes 
is there plan to rebase on #11592 and update this PR?

nielsbasjes commented:
Yes, this is the next thing on my todo list. Right now I'm preparing my talk for BerlinBuzzwords which is in a few weeks.

azagrebin commented:
ok, thanks for the update

nielsbasjes commented:
@azagrebin I did a rebase against the latest master, cleanup and implemented the dynamic docker image tag for the documentation in more places.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:
```diff
@@ -103,6 +103,15 @@ public static String getGitCommitTimeString() {
+	 *
+	 * @return The "name:tag" of the Flink docker image.
+	 */
+	public static String getFlinkDockerImageNameAndTag() {
```
Why  do you think this function belongs to `EnvironmentInformation `?
It looks to me as just a private helper method for `CONTAINER_IMAGE` option for now.

nielsbasjes replied:
Understood. I've moved it to the simplest implementation I can think of (i.e. only in the KubernetesConfigOptions ). If the real usecase appears I expect it to be more clear what the appropriate "central" location should be.

azagrebin commented on docs/ops/deployment/kubernetes.md:
```diff
@@ -159,6 +159,8 @@ with the `kubectl` command:
 ## Appendix
 
 ### Common cluster resource definitions
+The Deployment definitions use the pre-built image `flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}`{% else %}latest` (which contains the latest released Flink built against Scala 2.12){% endif %} which can be found [on Docker Hub](https://hub.docker.com/_/flink/).
+The image is built from this [Github repository](https://github.com/apache/flink-docker).
```
We already have a chapter on this page about the image, I would rather refer to it: `#flink-docker-image`.
This way we would have to change the image description only in one place. Also, the information about scala 2.12 for latest can become stale, it would make sense to just refer to the tags page. We have everywhere Scala 2.11 by default except docker and might want to make it 2.11 for docker as well later.

nielsbasjes replied:
Yes, I fixed this. I've added a bit in the central docker documentation files instead.

nielsbasjes commented on docs/ops/deployment/kubernetes.md:
```diff
@@ -262,7 +261,7 @@ spec:
       containers:
       - name: jobmanager
-        image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %}
+        image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest # This contains the latest released Flink built against Scala 2.12{% endif %}
```
Yes, I'll reword to make it better maintainable.

nielsbasjes replied:
I came up with the generic rewording (which only is shown if the documentation is built for a snapshot version which will show 'latest'):
``` The 'latest' tag contains the latest released version of Flink for a specific Scala version which will mismatch with your application over time.```

nielsbasjes commented on docs/ops/deployment/kubernetes.md:
```diff
@@ -159,7 +159,6 @@ with the `kubectl` command:
 ## Appendix
 
 ### Common cluster resource definitions
-
```
Yes

azagrebin commented on docs/ops/deployment/kubernetes.md:
```diff
@@ -262,7 +262,7 @@ spec:
       containers:
       - name: jobmanager
-        image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %}
+        image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest # The 'latest' tag contains the latest released version of Flink for a specific Scala version which will mismatch with your application over time.{% endif %}
```
```suggestion
        image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest # The 'latest' tag contains the latest released version of Flink for a specific Scala version which can mismatch with the Scala version of your application.{% endif %}
```
Will the Scala version definitely mismatch?
I am wondering why this comment should be in all places. This is for dev docs.
Could we not put this comment to that note in the relevant docker/tags chapter?

nielsbasjes replied:
In my experience: Yes, the versions will mismatch at some point. 
Also with a Scala mismatch the problem is big, yet with a Flink version mismatch I suspect more subtle problems to occur.

Also I consider the choice of having Scala 2.12 the version used for "latest" a choice for today which may change (As I mentioned in the mailing list I am actually against having a "latest" for something like Flink).

Note that this comment is ONLY shown when the documentation is generated for a non-stable "SNAPSHOT" build of the documentation. So the normal documentation should not show this.
I duplicated it in all of those places because these are Kubernetes config files I expect to be used by operations people who will have limited knowledge of the application at hand.

azagrebin replied:
Alright but if we and user always use the same Scala version, it will never mismatch, right?

nielsbasjes replied:
For me one big trigger was that I had a mis match and the errors were not helping me figure out what was happening.

azagrebin replied:
This sounds fine for me to fix.
It just sounds now as if it `will` definitely mismatch, although it looks like it `can` mismatch if not aligned with the image.

azagrebin replied:
How about this:
```
flink:latest # The 'latest' tag contains the latest released version of Flink for a specific Scala version which can conflict with the versions, used by your application.
```
Somehow this sounds confusing:
`which will mismatch with your application over time`, in particular `over time`.

zentol replied:
I think we should be way more explicit about this; scala users should _never_ use `latest`; it is simply to easy to run into issues. If we release a newer release, with scala 2.13 support we suddenly break everything.
I would explicitly phrase it like that; use it only if you are only using java, otherwise select a specific Flink version with a matching scala version.

We should also look into providing a `latest_<scala.version>` tag to make this easier for scala users.

azagrebin replied:
We added a [note](https://github.com/apache/flink/pull/11245/files#diff-a530b60989203f9e1f03c64c57deb56cR49) for this which can be promoted to warning.
I think the idea here is to keep the comment in this particular line (and others like this) brief enough as it is a code example.

nielsbasjes commented:
Did rebase to resolve conflicts

azagrebin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -137,11 +138,17 @@
+		.defaultValue(DEFAULT_CONTAINER_IMAGE)
+		.withDescription("Image to use for Flink containers. " +
+			"The specified image MUST be based upon Apache Flink " +
+			EnvironmentInformation.getVersion() + " that uses Scala " + EnvironmentInformation.getScalaVersion() + " . " +
```
Sorry, I did not notice this before.
This text goes into the user docs. The Scala version used by the build will be fixed in the user docs.
Therefore, It makes sense to require that the Scala version of the user application and the image must be the same.

btw, when any option is changed, the contract now is to [rebuild the option docs](https://github.com/apache/flink/tree/master/flink-docs#configuration-documentation).
This is why the CI fails.

I also think we do not usually highlight imperatives like `MUST` in user docs for style reasons. It looks a bit like shooting...

azagrebin replied:
I also think we do not usually highlight imperatives like MUST in user docs for style reasons. It looks a bit like shooting...

nielsbasjes replied:
Fixed the 'MUST'

nielsbasjes replied:
Thanks. I regenerated the documentation which now has the values of the default and the description with the values as they were on my machine when I generated it. I can make the description  'generic' but even then the default value will show "what I have".
So everytime someone changes any config documentation this default in the documentation will change to what they have.

How do you propose we handle this?

azagrebin replied:
This is a good point. The default value, generated like this, does not look like making sense then. I think the ideal solution would be to put a default value which includes only released Flink version or `latest` for snapshot, both without Scala version, mostly for documentation purposes.

The actual image to use in runtime, if this option is not explicitly configured by user and what we can document, can be then the current full image tag: Flink and Scala version which were used to build the running Flink. This requires modification of the code which parses the option value.

Ideally, we should also calculate the concrete latest Flink (e.g. 1.10) for snapshot as we do not have docker image for snapshot version (e.g. 1.11-snapshot). This is to combine it with the concrete Scala version, compiled against. For simplicity, we could ignore Scala version for snapshots for now and just use `latest`, like in docs, as it is mostly for development where we could ensure correct Scala version for testing.

azagrebin replied:
If you had problems with Flink/Scala version mismatch, maybe the problem is actually more general. The solution could be to embed some explicit runtime check of the versions used to build the user code and the cluster where the user tries to submit the job to. This way we could fail fast with a descriptive mismatch message or at least log the mismatch. I am not aware whether we have something like this. This also sounds like a separate effort.

nielsbasjes replied:
I was thinking about this and the main thing is handling version difference between the flink runtime and the application flink libraries better.

Now I see two things that can be done to avoid this:
1) Avoid: By setting the right default docker image to use in Kubernetes (this story). I think we should simply have a SNAPSHOT image too. I do this for my own projects too https://hub.docker.com/r/nielsbasjes/yauaa/tags 
2) Block: At runtime check the compatibility between the application and the runtime. I agree this is a separate problem. I propose to make a new issue out of that.

So this change is about the Avoid part, yet it turns out that there are for each Flink version at least 2 distinct versions that are incompatible. For the code this is fine as this is generated for each; the documentation is going to 'flip flop'.

So the real problem is that this default depends on the actual build using the specific tooling version of the developer. The problem is that this has more variations than the version of the project and this is reflected in the documentation.
So really the actual default and the default in the documentation should be different!
How can I do that? Because the table in the documentation should simply show "2.12" as the default scala version; yet the actual code should have the REAL scala version to be used as a default.

nielsbasjes replied:
By doing this default container image without the scala version would let the problem exist I intended to solve. Actually the version should "flip flop" with the scala it was built against, just not in the documentation.

So I thought of a different way to stabilized the version just for the documentation. It is a bit onorthodox so I'm curious to hear what you think of this direction.

The alternative direction I had in mind would be extending the ConfigOption with something like "defaultValueForDocumentation".

azagrebin replied:
What is the use case of the image's `ConfigOption#defaultValue` in the user code during runtime? A separate `ConfigOption#defaultValueForDocumentation` would make sense if there is a good reason that the user code needs the actual value with Scala version from `ConfigOption#defaultValue` during runtime. From what I see atm, Flink code does not necessarily need it.

Flink uses `ConfigOption#defaultValue` to generate web documentation which should not "flip flop" but Flink does not have to use `ConfigOption#defaultValue` for the actual default value during runtime. The actual default value can be derived by Flink code from Flink and Scala version (same as `ConfigOption#defaultValue` was implemented now) in runtime if user has not configured it explicitly (`configuration.contains("image.option") == false`).

nielsbasjes replied:
When running on K8s the jobmanager needs the exact image name at runtime to dynamically start new task workers. So the user application code does not need the image name, the Flink infra running on K8s does need it at runtime.

So what you propose is to have the defaultValue be only used by the documentation and the actual runtime default be calculated in a different place? I would vote against this. 
Making the way documentation is generated "obscure" (as I did in my last patch) is less of an issue to me than making the actually used runtime code "obscure".

nielsbasjes replied:
@azagrebin I changed it to the "defaultValueForDocumentation" construct as I mentioned and now it all looks so much cleaner and easier to use.
The documentation now shows a stable flink and scala version without "SNAPSHOT" and the actual default contains the exact value as is needed by the version at hand.

I also did a rebase onto the latest master.

azagrebin replied:
`defaultValueForDocumentation` now still differs from the actual runtime value. It can be still confusing that the default image has a concrete Scala version 2.12.

I have another suggestion. We have a bunch of options which have placeholders in the default values, e.g. `MetricOptions#REPORTER_CONFIG_PARAMETER`.
We could do the same here:
`flink-1.11-scala_<SCALA_VERSION>`
explaining, what the `<SCALA_VERSION>` is, in the option description.

azagrebin replied:
As for `defaultValueForDocumentation`, I just realised, we already have `org.apache.flink.annotation.docs.Documentation.OverrideDefault` annotation for this. Check e.g. `io.tmp.dirs` option. This looks like what we need.

nielsbasjes replied:
Yes, this annotation is better.
There is one downside: in an annotation text you cannot have any kind of dynamic parameter. So we can not even have the version of the project in there.
So I changed it essentially to a description of the default value similar to what was done in the case of the TMP_DIRS option.

nielsbasjes commented:
@azagrebin (Weird, cannot respond to your feedback regarding the comment with 'latest' directly...)

I put the "over time" mention in because your application may work today but as soon as a new version is released or Scala 2.13 is added it will break in unexpected ways.

How about this? ```# The 'latest' tag contains the latest released version of Flink for a specific Scala version. Do NOT use this in production as it will break your setup automatically when a new version is released.```

I still strongly believe we should simply NOT have the 'latest' tag at all and we should have docker images with SNAPSHOT versions which are pushed at the same time a SNAPSHOT build is pushed to maven. That would make this entire situation go away cleanly.

azagrebin commented:
@nielsbasjes
The last suggestion looks good to me. Let's just avoid uppercase: `NOT` -> `not`.
 I agree, snapshot versions would make things cleaner, hopefully community will get to it soon.

nielsbasjes commented:
Note that this should be squashed into a single commit before merging.

azagrebin commented:
The option docs has to be rebuilt.

nielsbasjes commented:
Side question: Where can I see the CI for Flink? 
This seems to be no longer used https://travis-ci.org/github/apache/flink/pull_requests ? 

I did a rebase onto the latest master and rebuilt the docs.

azagrebin commented:
You can see the CI status for the PR in a comment which is automatically created and updated by the Flink bot.
It is usually in the third comment, like for [this PR here](https://github.com/apache/flink/pull/11245#issuecomment-592002512).

Since 1.11, [the Flink CI](https://cwiki.apache.org/confluence/display/FLINK/2020/03/22/Migrating+Flink%27s+CI+Infrastructure+from+Travis+CI+to+Azure+Pipelines) is hosted by [Azure Pipelines](https://dev.azure.com/apache-flink/apache-flink/_build?definitionId=2), so it is not Travis anymore. 1.10 still uses Travis.

azagrebin commented:
@nielsbasjes I adjusted the commit message and added one more fixup commit.
The fixup commit uses `latest` tag for `snapshot` versions as we do not have `snapshot` Flink images yet.

nielsbasjes commented:
Using the 'latest' in case of snapshot is consistent with the documentation.
So LGTM     
+1
Commit when green.
